File: run_migration_db_updated_at.py
import urllib.request
import json

# .env.local から環境変数を読み込む
env = {}
with open('backend/.env.local', encoding='utf-8') as f:
    for line in f:
        line = line.strip()
        if '=' in line and not line.startswith('#'):
            k, v = line.split('=', 1)
            env[k.strip()] = v.strip().strip('"')

supabase_url = env.get('SUPABASE_URL', '').rstrip('/')
service_key = env.get('SUPABASE_SERVICE_KEY', '')

print(f'Supabase URL: {supabase_url[:40]}...')

# buyers テーブルに db_updated_at カラムが存在するか確認
# information_schema を直接クエリ
check_url = supabase_url + '/rest/v1/information_schema.columns?table_name=eq.buyers&column_name=eq.db_updated_at&select=column_name,data_type'

headers = {
    'apikey': service_key,
    'Authorization': f'Bearer {service_key}',
}

try:
    req = urllib.request.Request(check_url, headers=headers)
    with urllib.request.urlopen(req, timeout=15) as r:
        result = json.loads(r.read())
        if result:
            print('✅ db_updated_at カラムは既に存在します:', result)
        else:
            print('❌ db_updated_at カラムが存在しません')
            print('\n以下のSQLをSupabase SQL Editorで実行してください:')
            print('ALTER TABLE buyers ADD COLUMN IF NOT EXISTS db_updated_at TIMESTAMP WITH TIME ZONE;')
            print(f'\nSQL Editor: https://supabase.com/dashboard/project/krxhrbtlgfjzsseegaqq/sql/new')
except Exception as e:
    print(f'確認失敗: {e}')
    print('\n以下のSQLをSupabase SQL Editorで実行してください:')
    print('ALTER TABLE buyers ADD COLUMN IF NOT EXISTS db_updated_at TIMESTAMP WITH TIME ZONE;')
    print(f'\nSQL Editor: https://supabase.com/dashboard/project/krxhrbtlgfjzsseegaqq/sql/new')

File: backend/run-migration-db-updated-at.ts
import { createClient } from '@supabase/supabase-js';
import * as dotenv from 'dotenv';
import * as path from 'path';

dotenv.config({ path: path.join(__dirname, '.env.local') });

const supabase = createClient(
  process.env.SUPABASE_URL!,
  process.env.SUPABASE_SERVICE_KEY!
);

async function runMigration() {
  console.log('buyers テーブルに db_updated_at カラムを追加中...');

  const { error } = await supabase.rpc('exec_sql', {
    sql: 'ALTER TABLE buyers ADD COLUMN IF NOT EXISTS db_updated_at TIMESTAMP WITH TIME ZONE;'
  });

  if (error) {
    // rpc が使えない場合は直接 REST API で実行
    console.log('rpc 失敗、別の方法を試みます:', error.message);

    // Supabase の管理 API を使用
    const response = await fetch(
      `${process.env.SUPABASE_URL}/rest/v1/rpc/exec_sql`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'apikey': process.env.SUPABASE_SERVICE_KEY!,
          'Authorization': `Bearer ${process.env.SUPABASE_SERVICE_KEY!}`,
        },
        body: JSON.stringify({
          sql: 'ALTER TABLE buyers ADD COLUMN IF NOT EXISTS db_updated_at TIMESTAMP WITH TIME ZONE;'
        })
      }
    );
    
    if (!response.ok) {
      const text = await response.text();
      console.error('マイグレーション失敗:', text);
      console.log('\n手動でSupabase SQL Editorで以下を実行してください:');
      console.log('ALTER TABLE buyers ADD COLUMN IF NOT EXISTS db_updated_at TIMESTAMP WITH TIME ZONE;');
      return;
    }
  }

  // カラムが追加されたか確認
  const { data, error: checkError } = await supabase
    .from('buyers')
    .select('db_updated_at')
    .limit(1);

  if (checkError) {
    console.error('確認エラー:', checkError.message);
    console.log('\n手動でSupabase SQL Editorで以下を実行してください:');
    console.log('ALTER TABLE buyers ADD COLUMN IF NOT EXISTS db_updated_at TIMESTAMP WITH TIME ZONE;');
  } else {
    console.log('✅ db_updated_at カラムが正常に追加されました');
  }
}

runMigration().catch(console.error);
